src/next_array.rs

```diff
@@ -0,0 +1,64 @@
+use core::mem::MaybeUninit;
+use core::ptr;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N], // Invariant: arr[..len] is initialized.
+    len: usize,               // Invariant: len <= N.
+}
```

Comment:
Good point; I've updated the suggestion accordingly.

Reply:
Also I must ask, will anyone... ever.. run `cargo doc --document-private-items` to look at the (safety) invariants of the private members of this private type instead of just looking at the code?

Reply:
I use it quite often on the crates I maintain, and there's hardly a downside of writing an extra `/`.